applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -0,0 +1,603 @@
+        rResult[index + 2] = iNode.GetDof(ADJOINT_DISPLACEMENT_Z).EquationId();
+
+        rResult[index + 3] = iNode.GetDof(ADJOINT_ROTATION_X).EquationId();
+        rResult[index + 4] = iNode.GetDof(ADJOINT_ROTATION_Y).EquationId();
+        rResult[index + 5] = iNode.GetDof(ADJOINT_ROTATION_Z).EquationId();
+    }
+
+}
+
+void AdjointFiniteDifferencingBaseElement::GetDofList(DofsVectorType& rElementalDofList,
+    ProcessInfo& rCurrentProcessInfo) {
+
+    const SizeType num_dofs = 6 * GetGeometry().PointsNumber(); // 6 dofs per node
+    rElementalDofList.resize(0);
+    rElementalDofList.reserve(num_dofs);
+
+    GeometryType & geom = this->GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); i++)
+    {
+        NodeType & iNode = geom[i];
+
+        rElementalDofList.push_back(iNode.pGetDof(ADJOINT_DISPLACEMENT_X));
+        rElementalDofList.push_back(iNode.pGetDof(ADJOINT_DISPLACEMENT_Y));
+        rElementalDofList.push_back(iNode.pGetDof(ADJOINT_DISPLACEMENT_Z));
+
+        rElementalDofList.push_back(iNode.pGetDof(ADJOINT_ROTATION_X));
+        rElementalDofList.push_back(iNode.pGetDof(ADJOINT_ROTATION_Y));
+        rElementalDofList.push_back(iNode.pGetDof(ADJOINT_ROTATION_Z));
+    }
+}
+
+void AdjointFiniteDifferencingBaseElement::GetValuesVector(Vector& rValues, int Step) {
+
+    const SizeType num_dofs = 6 * GetGeometry().PointsNumber(); // 6 dofs per node
+    if(rValues.size() != num_dofs)
+        rValues.resize(num_dofs, false);
+
+    const GeometryType & geom = this->GetGeometry();
+    const int dimension = geom.WorkingSpaceDimension();
+
+    for (SizeType i = 0; i < geom.size(); i++)
+    {
+        const NodeType & iNode = geom[i];
+        const array_1d<double,3>& disp = iNode.FastGetSolutionStepValue(ADJOINT_DISPLACEMENT, Step);
+        const array_1d<double,3>& rot = iNode.FastGetSolutionStepValue(ADJOINT_ROTATION, Step);
+
+        const SizeType index = i * dimension * 2;
+        rValues[index]     = disp[0];
+        rValues[index + 1] = disp[1];
+        rValues[index + 2] = disp[2];
+
+        rValues[index + 3] = rot[0];
+        rValues[index + 4] = rot[1];
+        rValues[index + 5] = rot[2];
+    }
+}
+
+void AdjointFiniteDifferencingBaseElement::Calculate(const Variable<Matrix >& rVariable, Matrix& rOutput,
+                                                const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+
+    if(rVariable == STRESS_DISP_DERIV_ON_GP)
+    {
+            this->CalculateStressDisplacementDerivative(STRESS_ON_GP, rOutput, rCurrentProcessInfo);
+    }
+    else if(rVariable == STRESS_DISP_DERIV_ON_NODE)
+    {
+        this->CalculateStressDisplacementDerivative(STRESS_ON_NODE, rOutput, rCurrentProcessInfo);
+    }
+    else if(rVariable == STRESS_DESIGN_DERIVATIVE_ON_GP)
+    {
+        const std::string design_varible_name = this->GetValue( DESIGN_VARIABLE_NAME );
+
+        if (KratosComponents<Variable<double>>::Has(design_varible_name) == true)
```

Comment:
usually the `== true` is skipped for the Has method

Reply:
Done